kubernetes/metadata/pod.go

```diff
@@ -88,25 +91,32 @@ func (p *pod) GenerateK8s(obj kubernetes.Resource, opts ...FieldOptions) mapstr.
 
 	out := p.resource.GenerateK8s("pod", obj, opts...)
 
-	// check if Pod is handled by a ReplicaSet which is controlled by a Deployment
-	// same happens with CronJob vs Job. The hierarchy there is CronJob->Job->Pod
+	// check if Pod is handled by a ReplicaSet which is controlled by a Deployment.
+	// The hierarchy there is Deployment->ReplicaSet->Pod.
 	if p.addResourceMetadata.Deployment {
-		rsName, _ := out.GetValue("replicaset.name")
-		if rsName, ok := rsName.(string); ok {
-			dep := p.getRSDeployment(rsName, po.GetNamespace())
-			if dep != "" {
-				_, _ = out.Put("deployment.name", dep)
+		if p.replicaset != nil {
+			rsName, _ := out.GetValue("replicaset.name")
+			if rsName, ok := rsName.(string); ok {
+				meta := p.replicaset.GenerateFromName(po.Namespace + "/" + rsName)
+				deploymentName, _ := meta.GetValue("deployment.name")
+				if deploymentName != "" {
+					_, _ = out.Put("deployment.name", deploymentName)
+				}
 			}
 		}
 	}
 
-	// check if Pod is handled by a Job which is controlled by a CronJob
+	// check if Pod is handled by a Job which is controlled by a CronJob.
+	// The hierarchy there is CronJob->Job->Pod
 	if p.addResourceMetadata.CronJob {
-		jobName, _ := out.GetValue("job.name")
-		if jobName, ok := jobName.(string); ok {
-			dep := p.getCronjobOfJob(jobName, po.GetNamespace())
-			if dep != "" {
-				_, _ = out.Put("cronjob.name", dep)
+		if p.job != nil {
+			jobName, _ := out.GetValue("job.name")
+			if jobName, ok := jobName.(string); ok {
+				meta := p.job.GenerateFromName(po.Namespace + "/" + jobName)
+				cronjobName, _ := meta.GetValue("cronjob.name")
+				if cronjobName != "" {
+					_, _ = out.Put("cronjob.name", cronjobName)
```

Comment:
Wondering if we can put a debug message here on every put action.
And then for cases that we want to troubleshoot OOM issues we can enable debug and check what put actions might take place. 

@ChrsMark would it be valuable?

Reply:
That's a valid point on high level, however I don't think a `debug` message at this place would actually help us. Here we just retrieve an already stored value and also this message would occur many times for big number of Pods so this would be an overkill. Sth helpful would be to check the size of the local caches (Informers' cache) but this is quite out of scope I think on this PR. In any case have created [https://github.com/elastic/beats/issues/35591](https://github.com/elastic/beats/issues/35591) to evaluate further improvements.